skimage/restoration/inpaint.py

```diff
@@ -0,0 +1,168 @@
+from __future__ import print_function, division
+
+import numpy as np
+import skimage
+from scipy import sparse
+from scipy.sparse.linalg import spsolve
+
+
+def inpaint_biharmonic(img, mask):
+    """Inpaint masked points in image with biharmonic equations.
+
+    Parameters
+    ----------
+    img : 2-D np.array
+        Input image.
+    mask : 2-D np.array
+        Array of pixels to be inpainted. Have to be the same size as 'img'.
+        Unknown pixels has to be represented with 1, known - with 0.
+
+    Returns
+    -------
+    out : 2-D np.array
+        Input image with masked pixels inpainted.
+
+    Example
+    -------
+    >>> row = np.square(np.linspace(0, 1, 5))
+    >>> img = np.repeat(np.reshape(row, (1, 5)), 5, axis=0)
+    >>> mask = np.zeros_like(img)
+    >>> mask[2, 2:] = 1
+    >>> mask[1, 3:] = 1
+    >>> mask[0, 4:] = 1
+    >>> out = inpaint_biharmonic(img, mask)
+
+    References
+    ----------
+    Algorithm is based on:
+    .. [1]  N.S.Hoang, S.B.Damelin, "On surface completion and image inpainting
+            by biharmonic functions: numerical aspects",
+            http://www.ima.umn.edu/~damelin/biharmonic
+
+    Realization is based on:
+    .. [2]  John D'Errico,
+        http://www.mathworks.com/matlabcentral/fileexchange/4551-inpaint-nans,
+            method 3
+    """
+
+    if img.ndim != 2 or mask.ndim != 2:
+        raise ValueError('Only 2-dimensional arrays are supported')
+    if img.shape != mask.shape:
+        raise ValueError('Input arrays have to be the same shape')
+    if np.ma.isMaskedArray(img):
+        raise TypeError('Masked arrays are not supported')
+
+    # TODO: add sufficient conditions (if any)
+
+    img = skimage.img_as_float(img)
+    mask = mask.astype(np.bool)
+
+    out = np.copy(img)
+    out_h, out_w = out.shape
+    out_l = out.size
+
+    def _in_bounds(idx):
+        if len(idx) == 1:
+            return 0 <= idx <= out_l - 1
+        else:
+            return (0 <= idx[0] <= out_h - 1) and (0 <= idx[1] <= out_w - 1)
+
+    # Find indexes of masked points in flatten array
+    mask_mn = np.array(np.where(mask)).T
```

Comment:
As far as I use both addressing methods (both 1D and 2D indices) in the code overall, I'd prefer to better distinguish them